src/librustc_codegen_ssa/back/link.rs

```diff
@@ -968,7 +968,80 @@ pub fn print_native_static_libs(sess: &Session, all_native_libs: &[NativeLibrary
     }
 }
 
+// Because windows-gnu target is meant to be self-contained for pure Rust code it bundles
+// own mingw-w64 libraries. These libraries are often not compatible with mingw-w64
+// installed in the system. This breaks many cases where Rust is mixed with other languages
+// (e.g. *-sys crates).
+// We prefer system mingw-w64 libraries if they are available to avoid this issue.
+fn get_crt_libs_path(sess: &Session) -> Option<PathBuf> {
+    fn find_exe_in_path<P>(exe_name: P) -> Option<PathBuf>
+    where
+        P: AsRef<Path>,
+    {
+        env::var_os("PATH").and_then(|paths| {
+            env::split_paths(&paths)
+                .filter_map(|dir| {
+                    let full_path = dir.join(&exe_name);
+                    if full_path.is_file() {
+                        Some(fix_windows_verbatim_for_gcc(&full_path))
+                    } else {
+                        None
+                    }
+                })
+                .next()
+        })
+    }
+
+    use std::sync::Mutex;
+    lazy_static::lazy_static! {
+        static ref SYSTEM_LIBS: Mutex<Option<Result<PathBuf, ()>>> = Mutex::new(None);
+    }
+
+    let system_libs = SYSTEM_LIBS.lock().unwrap().clone();
+    match system_libs {
+        Some(Ok(compiler_libs_path)) => Some(compiler_libs_path),
+        Some(Err(_)) => None,
+        _ => {
+            if let (linker, LinkerFlavor::Gcc) = linker_and_flavor(&sess) {
+                let linker_path = if cfg!(windows) { linker.with_extension("exe") } else { linker };
+                if let Some(linker_path) = find_exe_in_path(linker_path) {
+                    let mingw_arch = match &sess.target.target.arch {
+                        x if x == "x86" => "i686",
+                        x => x,
+                    };
+                    let mingw_dir = [mingw_arch, "-w64-mingw32"].concat();
```

Comment:
I think it's probably more clear to use `format!("{}-...", mingw_arch)` here perhaps? (random stylistic nit)

Reply:
Unnecessary micro-optimisation 🤷‍♂